### Install depencies

In [1]:
!pip install unsloth bitsandbytes transformers accelerate
!pip install gradio
!pip install PyMuPDF  # for PDF reading

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.8/46.8 kB 3.5 MB/s eta 0:00:00
INFO: pip is looking at multiple versions of torchvision to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 218.5/218.5 kB 8.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.1/76.1 MB 10.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 491.5/491.5 kB 33.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 162.1/162.1 kB 16.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 318.9/318.9 kB 26.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.0/129.0 kB 13.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 31.5/31.5 MB 47.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 865.2/865.2 MB 1.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 156.5/156.5 MB 6.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 393.1/

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.1/54.1 MB 19.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 322.9/322.9 kB 31.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 95.2/95.2 kB 9.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.5/11.5 MB 61.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 72.0/72.0 kB 8.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.5/62.5 kB 6.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.0/20.0 MB 107.2 MB/s eta 0:00:00


#### Import packages

In [34]:
import unsloth
unsloth.__version__

'2025.4.7'

In [38]:
import accelerate
accelerate.__version__

'1.6.0'

In [35]:
import transformers
transformers.__version__

'4.51.3'

In [2]:
from unsloth import FastLanguageModel
from transformers import AutoTokenizer

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!


### Import model from Hugging Face

- Login into HuggingFace

In [3]:
from huggingface_hub import notebook_login
notebook_login()

- Import model

In [4]:
model_id = "aman012/mistral-7b-instruct-v0.3-bnb-4bit-200"
MAX_SEQ_LENGTH = 2048
dtype = None
load_in_4bit = True

In [6]:
model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = model_id,
    max_seq_length = MAX_SEQ_LENGTH,
    dtype = dtype,
    load_in_4bit = load_in_4bit,
)

==((====))==  Unsloth 2025.4.7: Fast Mistral patching. Transformers: 4.51.3.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.7.0+cu126. CUDA: 7.5. CUDA Toolkit: 12.6. Triton: 3.3.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.30. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


Unsloth 2025.4.7 patched 32 layers with 32 QKV layers, 32 O layers and 32 MLP layers.


In [7]:
model.eval()

PeftModelForCausalLM(
  (base_model): LoraModel(
    (model): MistralForCausalLM(
      (model): MistralModel(
        (embed_tokens): Embedding(32768, 4096, padding_idx=770)
        (layers): ModuleList(
          (0-31): 32 x MistralDecoderLayer(
            (self_attn): MistralAttention(
              (q_proj): lora.Linear4bit(
                (base_layer): Linear4bit(in_features=4096, out_features=4096, bias=False)
                (lora_dropout): ModuleDict(
                  (default): Identity()
                )
                (lora_A): ModuleDict(
                  (default): Linear(in_features=4096, out_features=16, bias=False)
                )
                (lora_B): ModuleDict(
                  (default): Linear(in_features=16, out_features=4096, bias=False)
                )
                (lora_embedding_A): ParameterDict()
                (lora_embedding_B): ParameterDict()
                (lora_magnitude_vector): ModuleDict()
              )
              (k_proj):

- Dummy Example

In [9]:
text = """Large Language Models(LLMs) are essential tools in natural language processing (NLP) and have been used in a variety of applications, such as text completion, translation, and question answering.

The output of large language models can be affected by various hyperparameters including temperature, top p, token length, max tokens and stop tokens.
Temperature

Temperature is a hyperparameter that controls the randomness of language model output.

A high temperature produces more unpredictable and creative results, while a low temperature produces more deterministic and conservative output. In other words, a higher temperature setting causes the model to be more “confident” in its output. A lower temperature setting yields more conservative and predictable output.

For example, if you adjust the temperature to 0.5, the model will generate text that is more predictable and less creative than if you set the temperature to 1.0.
Top p

Top p, also known as nucleus sampling, is another hyperparameter that controls the randomness of language model output.

It sets a threshold probability and selects the top tokens whose cumulative probability exceeds the threshold. The model then randomly samples from this set of tokens to generate output. This method can produce more diverse and interesting output than traditional methods that randomly sample the entire vocabulary.

For example, if you set top p to 0.9, the model will only consider the most likely words that make up 90% of the probability mass.
Token length

This is the number of words or characters in a sequence or text that is fed to the LLM.

It varies depending on the language and the tokenization method used for the particular LLM.

The length of the input text affects the output of the LLM.

A very short input may not have enough context to generate a meaningful completion.

Conversely, a rather long input may make the model inefficiently process or it may cause the model to generate an irrelevant output.
Max tokens

This is the maximum number of tokens that the LLM generates.

Within this, is the token limit; the maximum number of tokens that can be used in the prompt and the completion of the model. Determined by the architecture of the model LLM, it refers to the maximum tokens that can be processed at once.

The computational cost and the memory requirements are directly proportional to the max tokens. Set a longer max token, and you will have greater context and coherent output text. Set a shorter max token, and you will use less memory and have a faster response but your output is prone to errors and inconsistencies.

During the training and fine-tuning of the LLM, the max token is set.

Contrary to fine-tuning token length during the generation of output, the coherence and length of the output is carefully set at inception, based on the specific task & requirements, without affecting other parameters that will likely need adjusting.
Stop tokens

In simple terms, it is the length of the output or response of an LLM.

So it signifies the end of a sequence in terms of either a paragraph or a sentence.

Similar to max tokens, the inference budget is reduced when the stop tokens are set low.

For example, when the stop tokens are set at 2, the generated text or output will be limited to a paragraph. If the stop tokens is set at 1, the generated text will be limited to a sentence.
Other relevant hyperparameters

There are many other hyperparameters that can affect large language model performance, such as learning rate, number of layers, hidden size, and dropout rate. These hyperparameters can impact the model’s training time, accuracy, and generalizability."""

### Define functions

- Extract text from pdf function

In [37]:
import fitz  # first install fPyMuPDF
#Extract text from pdf
print(fitz.__version__)
def extract_text_from_pdf(file):
    doc = fitz.open(file)
    text = ""
    for page in doc:
        text += page.get_text()
    return text


1.25.5


- Split text into chunks function

We have trained out model with max_token 32768, number_of_token > 2048 will exceeds model's maximum sequence length. So we have to divide whole pdf into small chunks

In [11]:
def split_text_to_chunks(text: str, max_token: int = 1600):
    tokens = tokenizer.tokenize(text) #split text into tokens
    chunks = []
    for i in range(0, len(tokens), max_token):
        chunk = tokenizer.convert_tokens_to_string(tokens[i:i+max_token])
        chunks.append(chunk)
    return chunks

- Summarize text function

In [31]:
import torch
torch.__version__

'2.7.0+cu126'

In [32]:
def summarize_text(text: str, summary_len: int, tempt: float, top_p: float):
    if not text or not text.strip():
        return "❗ Please enter some text before summarizing."
    try:
        prompt = f"Summarize the following text into a concise paragraph of about {summary_len} words:\n\n{text}\n\nSummary:"
        inputs = tokenizer(prompt, return_tensors='pt').to('cuda')
        with torch.no_grad():
            outputs = model.generate(
                **inputs,
                max_new_tokens = summary_len * 2,
                temperature = tempt,
                do_sample = False,
                top_p = top_p
            )

            Summary = tokenizer.decode(outputs[0], skip_special_tokens=True)
            return Summary.replace(prompt, "").strip()
    except Exception as e:
        return f"An error occurred during summarization: {str(e)}"

In [13]:
summarize_text(text, 30, 0.9, 0.9)

'Large Language Models(LLMs) are essential tools in natural language processing (NLP) and have been used in a variety of applications, such as text completion, translation, and question answering.\n\nThe output of large language models can be affected by various hyperparameters including temperature,'

- Summarize pdf function

In [36]:
import tqdm
tqdm.__version__

'4.67.1'

In [21]:
#Summarization
from typing import Any
from tqdm import tqdm #use to see the progress bar of every chunk
def summarize_pdf(file: Any, summary_len: int, tempt: float, top_p: float):
    pdf_text = extract_text_from_pdf(file)
    if not pdf_text or not pdf_text.strip():
        return "Please upload any pdf first !"
    try:
        prompt = f"Summarize the following text into a concise paragraph of about {summary_len} words:\n\n{pdf_text}\n\nSummary:"

        # split into chunks
        chunks = split_text_to_chunks(pdf_text)

        #summarize each chunks
        summary_chunks = []
        for chunk in tqdm(chunks,  desc="Summarizing chunks"):
            summary = summarize_text(chunk, summary_len, tempt, top_p)
            summary_chunks.append(summary)

        # combine all summaries
        Summary = ' '.join(summary_chunks)
        final_summary = summarize_text(Summary, summary_len, tempt, top_p)
        return final_summary
    except Exception as e:
        return f"An error occurred during summarization: {str(e)}"

In [22]:
summarize_pdf('/content/Living in the light - a guide to personal and planetary - Shakti Gawain, with Laurel King - ( WeLib.org )-20-40.pdf', 200, 0.8, 0.5)

Summarizing chunks: 100%|██████████| 5/5 [01:40<00:00, 20.17s/it]


'Shakti Gawain was born in 1948 in New York. She was raised in a family that was not religious. She was always interested in spirituality and personal development. She moved to the San Francisco Bay area, which was the forefront of the so-called “human potential movement.” She took classes and workshops, read new books, meditated, and talked constantly to others involved in the same process. After reading Handbook to Higher Consciousness by Ken Keyes, she went to live at his center in Berkeley where they worked on their growth intensely, day and night, for a year. After that, she continued to live communally for several years with others who were involved in an intensive personal development process. During this time, she did whatever she could to make enough money to live on — housework, office work, odds and ends — while she focused on her real work. Since that time, over twenty-five years ago, her life has been dedicated to her growth and evolution as a conscious being. She graduall

#### Build a web interface using Gradio

In [27]:
import gradio
gradio.__version__

'5.29.0'

In [23]:
import gradio as gr
with gr.Blocks() as demo:
    gr.Markdown("## 🧠 Smart PDF/Text Summarizer with LLM\nUpload a file or type text. Customize your summary!")
    with gr.Tab("Type text"):
        text_input = gr.Textbox(lines=10, placeholder="Input your text here", label="Text Input")
        text_output = gr.Textbox(max_lines=100, placeholder="Show your output here", show_copy_button=True, label="Output")
        summary_len1 = gr.Slider(20, 200, value=30, step=1, label="length")
        temperature1 = gr.Slider(0.1, 1.5, value=0.8, step=0.1, label="Temperature")
        top_p1 = gr.Slider(0.1, 1.0, value=0.8, step=0.1, label="Top-p (nucleus sampling)")
        text_btn = gr.Button("Summarize Text")
        text_btn.click(fn=summarize_text, inputs=[text_input, summary_len1, temperature1, top_p1], outputs=text_output)
    with gr.Tab("upload pdf"):
        pdf_input = gr.File(label="Upload PDF")
        pdf_output = gr.Textbox(max_lines=100, placeholder="Show your output here", show_copy_button=True, label="Output")
        summary_len2 = gr.Slider(20, 200, value=30, step=1, label="length")
        temperature2 = gr.Slider(0.4, 1.3, value=0.7, step=0.1, label="Temperature")
        top_p2 = gr.Slider(0.6, 1.0, value=0.5, step=0.1, label="Top-p (nucleus sampling)")
        pdf_output = gr.Textbox(max_lines=100, placeholder="Show your output here", show_copy_button=True, label="Output")
        pdf_btn = gr.Button("Summarize PDF")
        pdf_btn.click(fn=summarize_pdf, inputs= [pdf_input, summary_len2, temperature2, top_p2], outputs=pdf_output)

    demo.launch(share = True, debug = True)

Colab notebook detected. This cell will run indefinitely so that you can see errors and logs. To turn off, set debug=False in launch().
* Running on public URL: https://4a954b3fd6692418e2.gradio.live

This share link expires in 1 week. For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)


Summarizing chunks: 100%|██████████| 2/2 [00:26<00:00, 13.48s/it]


Keyboard interruption in main thread... closing server.
Killing tunnel 127.0.0.1:7860 <> https://4a954b3fd6692418e2.gradio.live
